test/k8s/res_limits_test.go

```diff
@@ -0,0 +1,512 @@
+
+						y, err := yaml.Marshal(pod)
+						if err != nil {
+							y = []byte(fmt.Sprintf("failed to marshal pod yaml: %v", err))
+						}
+						errCh <- fmt.Errorf("pod '%s' failed in phase '%s'; pod yaml: %s, pod logs: %s",
+							pod.Name, pod.Status.Phase, y, podLogs)
+						watcher.Stop()
+						return
+					case <-ctx2.Done():
+						watcher.Stop()
+						close(errCh)
+						return
+					}
+				}
+			}()
+		}
+
+		BeforeAll(func() {
+			cpuLimit := requireVar("PERF_LIMIT_MILLI_CPU")
+			i, err := strconv.Atoi(cpuLimit)
+			Expect(err).ToNot(HaveOccurred(), "invalid value of PERF_LIMIT_MILLI_CPU")
+			cpu = i
+
+			memLimit := requireVar("PERF_LIMIT_MEGA_MEMORY")
+			i, err = strconv.Atoi(memLimit)
+			Expect(err).ToNot(HaveOccurred(), "invalid value of PERF_LIMIT_MEGA_MEMORY")
+			memory = i
+
+			svcNum, _ := os.LookupEnv("PERF_TEST_NUM_SERVICES")
+			insNum, _ := os.LookupEnv("PERF_TEST_INSTANCES_PER_SERVICE")
+
+			if svcNum != "" {
+				numServices, _ = strconv.Atoi(svcNum)
+			}
+			if insNum != "" {
+				instancesPerService, _ = strconv.Atoi(insNum)
+			}
+		})
+
+		BeforeEach(func() {
+			Expect(cluster.Install(NamespaceWithSidecarInjection(TestNamespace))).To(Succeed())
+		})
+
+		AfterEach(func() {
+			Expect(cluster.DeleteNamespace(TestNamespace)).To(Succeed())
+		})
+
+		// should not crash, dump memory usage
+		It("should deploy all services and instances", func() {
+
+			adjustResource(cpu, memory, false)
+
+			deployDPs()
+
+			errCh := make(chan error)
+			ctx, cancelMonitoring := context.WithCancel(context.Background())
+			watchControlPlane(ctx, errCh)
+
+			dpCh := make(chan bool)
+			waitForDPs(dpCh)
+
+			var err error
+			select {
+			case dpRet := <-dpCh:
+				printLog("dpCh returned\n")
+				Expect(dpRet).To(BeTrue(), "data planes should be run and available")
+				break
+			case err = <-errCh:
+				printLog("errCh returned\n")
+				break
+			}
+			cancelMonitoring()
+			Expect(err).ToNot(HaveOccurred(), "control plane should not crash")
+		})
+
+		// half the resource, should crash without GOMEMLIMIT
+		It("should crash when deploy all services and instances with half CP resource", func() {
+			adjustResource(cpu/2, memory/2, false)
+
+			deployDPs()
+
+			errCh := make(chan error)
+			ctx, cancelMonitoring := context.WithCancel(context.Background())
+			watchControlPlane(ctx, errCh)
+
+			dpCh := make(chan bool)
+			waitForDPs(dpCh)
+
+			var err error
+			select {
+			case <-dpCh:
+				printLog("dpCh returned\n")
+				break
+			case err = <-errCh:
+				printLog("errCh returned\n")
+				break
+			}
+			cancelMonitoring()
+			Expect(err).To(HaveOccurred(), "control plane should crash with half resource")
+			if err != nil {
+				_, _ = fmt.Fprint(GinkgoWriter, err.Error())
+			}
+
+			cpPods, err := k8s.ListPodsE(cluster.GetTesting(), cluster.GetKubectlOptions(Config.KumaNamespace), metav1.ListOptions{
+				LabelSelector: fmt.Sprintf("app=%s", Config.KumaServiceName),
+			})
```

Comment:
how about `cluster.GetKuma().GetKumaCPPods`

Reply:
Here is just to print container statuses of failed pods. So not logs. Logs are captured and printed in the watching.